File: solutions/typescript/alphametics/1/alphametics.ts
export function solve(puzzle: string): Record<string, number> | undefined {
  // Parse the puzzle: split into left side and right side
  const [left, right] = puzzle.split(' == ');
  const leftWords = left.split(' + ');
  const rightWord = right;
  
  // Extract all unique letters from the puzzle
  const allLetters = new Set<string>();
  leftWords.forEach(word => word.split('').forEach(char => allLetters.add(char)));
  rightWord.split('').forEach(char => allLetters.add(char));
  const letters = Array.from(allLetters);
  
  // Identify leading letters (cannot be zero)
  const leadingLetters = new Set<string>();
  leftWords.forEach(word => leadingLetters.add(word[0]));
  leadingLetters.add(rightWord[0]);
  
  // Helper function to convert word to number using current mapping
  function wordToNumber(word: string, mapping: Map<string, number>): number {
    let num = 0;
    for (const char of word) {
      num = num * 10 + mapping.get(char)!;
    }
    return num;
  }
  
  // Helper function to check if current mapping is valid
  function isValid(mapping: Map<string, number>): boolean {
    // Check leading digits are not zero
    for (const letter of leadingLetters) {
      if (mapping.get(letter) === 0) {
        return false;
      }
    }
    
    // Calculate sum of left side
    let sum = 0;
    for (const word of leftWords) {
      sum += wordToNumber(word, mapping);
    }
    
    // Calculate right side
    const result = wordToNumber(rightWord, mapping);
    
    return sum === result;
  }
  
  // Generate all permutations of digits for the letters
  const digits = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
  const usedDigits = new Set<number>();
  const mapping = new Map<string, number>();
  
  function backtrack(index: number): boolean {
    if (index === letters.length) {
      return isValid(mapping);
    }
    
    const letter = letters[index];
    for (const digit of digits) {
      if (usedDigits.has(digit)) continue;
      if (digit === 0 && leadingLetters.has(letter)) continue;
      
      usedDigits.add(digit);
      mapping.set(letter, digit);
      
      if (backtrack(index + 1)) {
        return true;
      }
      
      usedDigits.delete(digit);
      mapping.delete(letter);
    }
    
    return false;
  }
  
  if (backtrack(0)) {
    const result: Record<string, number> = {};
    for (const [letter, digit] of mapping) {
      result[letter] = digit;
    }
    return result;
  }
  
  return undefined;
}

File: solutions/java/alphametics/1/src/main/java/Alphametics.java
import java.util.*;

class Alphametics {
    private final List<String> leftWords;
    private final String rightWord;
    private final List<Character> uniqueChars;
    private final Set<Character> leadingChars;
    private final Map<Character, Integer> charToIndex;
    private final long[] charWeights;
    private Map<Character, Integer> solution;

    Alphametics(String userInput) {
        String[] parts = userInput.split("==");
        String leftSide = parts[0].trim();
        String rightSide = parts[1].trim();
        
        this.leftWords = new ArrayList<>();
        for (String word : leftSide.split("\\+")) {
            this.leftWords.add(word.trim());
        }
        this.rightWord = rightSide;
        
        this.uniqueChars = new ArrayList<>();
        this.leadingChars = new HashSet<>();
        this.charToIndex = new HashMap<>();
        
        Set<Character> seen = new HashSet<>();
        
        // Process left side words
        for (String word : leftWords) {
            if (word.length() > 1) {
                leadingChars.add(word.charAt(0));
            }
            for (char c : word.toCharArray()) {
                if (seen.add(c)) {
                    uniqueChars.add(c);
                    charToIndex.put(c, uniqueChars.size() - 1);
                }
            }
        }
        
        // Process right side word
        if (rightWord.length() > 1) {
            leadingChars.add(rightWord.charAt(0));
        }
        for (char c : rightWord.toCharArray()) {
            if (seen.add(c)) {
                uniqueChars.add(c);
                charToIndex.put(c, uniqueChars.size() - 1);
            }
        }
        
        // Precompute the mathematical "weight" of each character
        // This allows O(N) validation instead of string parsing
        this.charWeights = new long[uniqueChars.size()];
        for (String word : leftWords) {
            long multiplier = 1;
            for (int i = word.length() - 1; i >= 0; i--) {
                charWeights[charToIndex.get(word.charAt(i))] += multiplier;
                multiplier *= 10;
            }
        }
        
        long multiplier = 1;
        for (int i = rightWord.length() - 1; i >= 0; i--) {
            charWeights[charToIndex.get(rightWord.charAt(i))] -= multiplier;
            multiplier *= 10;
        }
    }

    Map<Character, Integer> solve() throws UnsolvablePuzzleException {
        // A base-10 system can only represent 10 unique digits
        if (uniqueChars.size() > 10) {
            throw new UnsolvablePuzzleException();
        }
        
        int[] assignment = new int[uniqueChars.size()];
        boolean[] used = new boolean[10];
        
        if (backtrack(0, assignment, used)) {
            return solution;
        } else {
            throw new UnsolvablePuzzleException();
        }
    }
    
    private boolean backtrack(int index, int[] assignment, boolean[] used) {
        // Base case: all characters have been assigned a digit
        if (index == uniqueChars.size()) {
            if (isValid(assignment)) {
                solution = new HashMap<>();
                for (int i = 0; i < uniqueChars.size(); i++) {
                    solution.put(uniqueChars.get(i), assignment[i]);
                }
                return true;
            }
            return false;
        }
        
        char c = uniqueChars.get(index);
        boolean isLeading = leadingChars.contains(c);
        
        // Try assigning digits 0-9
        for (int d = 0; d <= 9; d++) {
            if (!used[d]) {
                // Leading digits of multi-digit numbers cannot be 0
                if (d == 0 && isLeading) {
                    continue;
                }
                
                used[d] = true;
                assignment[index] = d;
                
                if (backtrack(index + 1, assignment, used)) {
                    return true;
                }
                
                used[d] = false; // Backtrack
            }
        }
        return false;
    }
    
    private boolean isValid(int[] assignment) {
        long sum = 0;
        for (int i = 0; i < uniqueChars.size(); i++) {
            sum += assignment[i] * charWeights[i];
        }
        return sum == 0;
    }
}